passes/trees.nim

```diff
@@ -0,0 +1,173 @@
+## Implements the ``PackedTree`` storage type, for space efficient storage of
+## arbitrary trees.
+
+import
+  std/strutils,
+  experimental/sexp
+
+type
+  TreeNode*[T] = object
+    ## A single node. How the val field is interpreted depens on the kind. For
+    ## leaf nodes, the field's meaning is decided externally, while for non-
+    ## leaf nodes, `val` stores how many child nodes the node has.
+    kind*: T
+    val*: uint32
+
+  PackedTree*[T] = object
+    ## Stores a node tree packed together in a single sequence.
+    nodes: seq[TreeNode[T]]
+    numbers: seq[uint64]
+    # TODO: use a BiTable for the numbers
+
+  NodeIndex* = distinct uint32
+
+const
+  ExternalFlag = 0x8000_0000'u32
+
+proc `[]`*[T](t: PackedTree[T], at: NodeIndex): TreeNode[T] {.inline.} =
+  t.nodes[ord at]
+
+proc next*(t: PackedTree, i: NodeIndex): NodeIndex =
+  ## Returns the index of the first node following `i` that is not a child
+  ## node of `i`.
+  mixin isAtom
+  var
+    i = uint32(i)
+    last = i
+  while i <= last:
+    if not isAtom(t.nodes[i].kind):
+      last += t.nodes[i].val
+    inc i
+
+  result = NodeIndex(i)
+
+proc child*(t: PackedTree, i: NodeIndex, child: Natural): NodeIndex =
+  ## Returns the index of the `child`-th subnode of `i`.
+  var i = uint32(i) + 1
+  for j in 0..<child:
```

Comment:
```suggestion
  for _ in 0..<child:
```

Only because it's unused, at first I thought the loop body had a bug because it wasn't using `j`, but I glossed over the `var I` declaration above; this change might make things a bit more clear.

Reply:
Yep, that's better.